Upstream navigation hooks from CSP

mikewest opened the pull request:
This will enable implementation of 'frame-ancestors' and 'form-action',
and makes a bit of progress towards whatwg/html#1230.

mikewest commented:
@annevk, this is probably something you need to sign off on to make sure the approach is sane.

mikewest commented:
Actually, hold off a moment. I need to make this a little bit more complicated. :)

annevk commented on source:
```diff
@@ -81628,6 +81630,11 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   <li><p>Let <var>navigationType</var> be "<code data-x="">form-submission</code>" if the <span
+   data-x="navigate">navigation algorithm</span> was invoked as a result of the <span
+   data-x="concept-form-submit">form submission algorithm</span>, and "<code data-x="">other</code>"
+   otherwise.</p></li>
```
At some point I'd like to get rid of this action-at-a-distance. It might be okay for now...

mikewest replied:
Yeah. I was surprised to see it, but since it's there... I was lazy. :(

annevk commented on source:
```diff
@@ -81639,7 +81646,8 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
      <dt>If <var>resource</var> is a <span data-x="concept-response">response</span></dt>
-     <dd><p>Run <span>process a navigate response</span> with <var>resource</var> and
+     <dd><p>Run <span>process a navigate response</span> with <var>resource</var>,
+     <var>navigationType</var>, the <span>source browsing context</span>, and
```
I guess you should pass _resource_ twice here? Or perhaps null as the request? Currently it doesn't really work.

mikewest replied:
Yeah. I didn't notice that... Hrm. I'll take another look at what |resource| is and whether it might work.

mikewest replied:
I think `null` works, but honestly I'm not sure when we hit this. In what cases do we navigate directly to a `response`?

mikewest replied:
I changed this to pass in `null` in https://github.com/whatwg/html/pull/1619/commits/679ab1c5ffa4e2b3908ea8ea3cd08c1e8ea85afd, but it's still not clear to me when this is called. I guess this is the only outstanding question on the PR?

annevk commented on source:
```diff
@@ -81920,20 +81942,35 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+     <li><var>response</var> is a network error</li>
+     <li><p class="&#x0058;&#x0058;&#x0058;"><var>response</var>'s <span data-x="concept-response-header-list">header
+     list</span> contains a header named `<code data-x="">X-Frame-Options</code>` whose value
+     expresses constraints which are violated by <var>browsingContext</var> (TODO: Spell this out)
```
Is the plan to leave this TODO in for now?

mikewest replied:
Yeah. One thing at a time. I need to look at the RFC to see if there's anything we can hook into, or we need to spell out all the behavior in this spec.

annevk commented:
I missed your question. We navigate to a response in `<object>`, `<embed>`, with `about:srcdoc`, and I think with `document.open()`.

mikewest commented:
> I missed your question. We navigate to a response in `<object>`, `<embed>`, with `about:srcdoc`, and I think with `document.open()`.

`null` seems reasonable for `about:srcdoc` and `document.open()`. I'm surprised that it happens for `<object>` and `<embed>`, but I don't think it affects any of the current usage for this new hook. XFO and `frame-ancestors` apply to things loaded via `<object>`/`<embed>` but that only requires looking at the response. The only thing currently looking at the request is `form-action`, which doesn't come into play when there's not a request to process.

If you're OK with the rest of the patch, I think it still seems reasonable.

annevk commented:
The way `<object data>` and `<embed src>` is that they fetch the URL first and then navigate to the response if they end up creating a browsing context for the response. Otherwise we'd have to fetch the data twice. Once there is a browsing context you can navigate normally in it of course, but those attributes would still behave in this way.

annevk commented on source:
```diff
@@ -81830,8 +81839,8 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
-  data-x="concept-request">request</span> <var>request</var> and <span>browsing context</span>
-  <var>browsingContext</var>, run these steps:</p>
+  data-x="concept-request">request</span> <var>request</var>, <span>browsing context</span>
+  <var>browsingContext</var>, and string <var>type</var>, run these steps:</p>
```
We should pass this "source browsing context" as well. Just like you do for "process a navigate response".

annevk replied:
Or would that get too messy further down? Could leave that as a TODO I suppose.

mikewest replied:
We should probably be passing it into "navigate", honestly. But that sounds like work. :)

Passing it into "process a navigate fetch" as of b357bbb.

mikewest commented:
I see. That makes sense.

annevk commented on source:
```diff
@@ -81959,20 +81981,35 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+     <li><p class="&#x0058;&#x0058;&#x0058;"><var>response</var>'s <span data-x="concept-response-header-list">header
+     list</span> contains a header named `<code data-x="">X-Frame-Options</code>` whose value
+     expresses constraints which are violated by <var>browsingContext</var> (TODO: Spell this out)
+     <ref spec="RFC7034"></p></li>
```
This RFC needs to be added to the references.

mikewest commented:
XFO needs more work; reduced it down to a bare TODO so that it's clear where the processing should happen in the future, but we'll need to actually spell it out here, as there's nothing in the RFC that we can easily point off to (perhaps @randomdross can take a stab at an HTMLized algorithm?).

Assigning @domenic to sign off on the final set of changes, per @annevk's comments in IRC: http://logs.glob.uno/?c=freenode%23whatwg#c1002867

mikewest commented:
Ping, @domenic?

domenic commented:
In meetings all day today, can try to do tomorrow...

domenic commented on source:
```diff
@@ -81876,14 +81886,15 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
   <ol>
    <li><p>Let <var>response</var> be null.</p></li>
 
    <li><p>Set <var>request</var>'s <span data-x="concept-request-client">client</span> to the
```
s/the // now

mikewest replied:
Dropped.

domenic commented on source:
```diff
@@ -82005,20 +82030,33 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+    <code>Document</code> object's <span>origin</span> set to a new <span
+    data-x="concept-origin-opaque">opaque origin</span>, and abort these steps.</p>
+
+    <ul class="brief">
```
I think this might look better without `class="brief"`

mikewest replied:
Removed.

domenic commented on source:
```diff
@@ -82005,20 +82030,33 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+    <ul class="brief">
+     <li><var>response</var> is a network error</li>
+     <li><p class="&#x0058;&#x0058;&#x0058;">TODO: Define <code data-x="">X-Frame-Options</code>
+     processing here.</p></li>
```
Maybe add a link to https://github.com/whatwg/html/issues/1230?

domenic commented on source:
```diff
@@ -82005,20 +82030,33 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+     processing here.</p></li>
+     <li>The <span>Should navigation response to navigation request of type from source in target be
+     blocked by Content Security Policy?</span> algorithm returns "<code data-x="">Blocked</code>"
+     when executed upon <var>request</var>, <var>response</var>, <var>type</var>, and
```
"type, source, and browsingContext" not "type, and source and browsingContext"

mikewest replied:
Good eye!

domenic commented on source:
```diff
@@ -81898,21 +81909,34 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
    <li><p>If <var>request</var>'s <span data-x="concept-request-method">method</span> is not `<code
```
Should we update other locations (like this one) that talk about "if the navigation algorithm was invoked as a result of the form submission algorithm" to instead check if _navigationType_ is "`form-submission`"? That seems like it would be nice.

mikewest replied:
The other usage is inside "process a navigate fetch", so we'd need to extend the argument list there; not the end of the world, I suppose. I didn't do it since we're pulling this value out of thin air anyway, but it probably makes sense to be explicit when we can.

mikewest replied:
Oh, no. I'm just wrong. We already pass it down, so you're absolutely right that we should use it there.

The term is also used in "plan to navigate", but I'm not sure it's worth the effort to pipe it through to there in this patch.

domenic commented:
Kind of sad the arguments lists grew so much but I guess there's not much you can do there. Found some editorial things worth tightening up.

domenic commented:
Can all of the CSP algorithms handle being called with a null request? It seems like you pass that through in a few circumstances.

mikewest commented:
Rebased and addressed your comments, @domenic. Thanks!

> Can all of the CSP algorithms handle being called with a null request? It seems like you pass that through in a few circumstances.

If y'all are happy with the approach in this patch, I'll update the CSP algorithms to exit early on a `null` request when relevant. Should be a trivial change.

annevk commented on source:
```diff
@@ -82185,20 +82208,33 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+    data-x="concept-origin-opaque">opaque origin</span>, and abort these steps.</p>
+
+    <ul>
+     <li><var>response</var> is a network error</li>
```
Use `<p>` here as well.

annevk replied:
Also needs a dot (or no dots anywhere in the list I suppose).

mikewest replied:
Done, thanks!

annevk commented on source:
```diff
@@ -82185,20 +82208,33 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+     <li>The <span>Should navigation response to navigation request of type from source in target be
+     blocked by Content Security Policy?</span> algorithm returns "<code data-x="">Blocked</code>"
+     when executed upon <var>request</var>, <var>response</var>, <var>type</var>, <var>source</var>,
+     and <var>browsingContext</var> <ref spec="CSP"></li>
```
Closing `</p>` missing.

annevk replied:
Also needs a dot.

mikewest replied:
Done and done.

annevk commented on source:
```diff
@@ -82185,20 +82208,33 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+    <ul>
+     <li><var>response</var> is a network error</li>
+     <li><p class="&#x0058;&#x0058;&#x0058;">TODO: Define <code data-x="">X-Frame-Options</code>
+     processing here [<a href="https://github.com/whatwg/html/issue/1230">whatwg/html#1230</a>].</p>
```
Closing `</li>` missing.

mikewest replied:
Done.

annevk commented on source:
```diff
@@ -82185,20 +82208,33 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+     blocked by Content Security Policy?</span> algorithm returns "<code data-x="">Blocked</code>"
+     when executed upon <var>request</var>, <var>response</var>, <var>type</var>, <var>source</var>,
+     and <var>browsingContext</var> <ref spec="CSP"></li>
+    </ul>
```
Newlines between the `<li>`s would be good too.

mikewest replied:
Done. No newlines between the `<ul>` and `<li>`, though, right?
